### ISSUE TITLE
Officially support a small repository of WORKSPACE VCS rules

### ISSUE THREAD

**Issue by pmcalpine** [bazelbuild/bazel#3304]
### TL;DR, please:

1. move the Skylark VCS fetch rules from `//tools/build_deps/repo` into their own,
officially supported, repository. Suggestion: github.com/bazelbuild/vcs_rules
2. deprecate the native WORKSPACE VCS rules in favour of this officially supported
vcs_rules repository.

My motivating objective in this request is to be able to efficiently override
how Bazel fetches dependencies without modifying the WORKSPACE file. My
intention/hope is to write my own implementations of the Skylark VCS fetch rules
and override the standard ones using the `--override_repository` command line argument.

### Problem: The native WORKSPACE rules cannot be overridden without modifying the WORKSPACE file and load(...)ing a replacement.

I think this essentially boils down to the native rules written in Java can't be
overridden with `--override_repository`. Therefore, I would ask that the native rules be
officially deprecated in favour of pure-Skylark rules, such as the ones in
`//tools/build_defs/repo`.

### Problem: Lack of support for VCS systems such as SVN, Mercurial, and Bazaar

Looking through `//tools/build_deps/repo` and
`//src/main/java/com/google/devtools/build/lib/bazel/rules/workspace` I only see
rules for HTTP, Git, and Maven. Fetching from other VCS systems requires a user to
implement their own rules. The result is that dependency fetching is done
inconsistently, making it more complicated to override how the dependencies are
fetched.

As an example, Go supports fetching from Git, SVN, Mercurial, and Bazaar. In
order to support these VCS systems the [rules_go](https://github.com/bazelbuild/rules_go) package has written its own
`fetch_repo` binary which uses the [golang.org/x/tools/go/vcs](https://godoc.org/golang.org/x/tools/go/vcs) package which is the same VCS
fetch code used by `go get`.

Even if all I were interested in were Git repositories I now have two places
where I need to override how git repositories may be fetched:

1. the `git_repository` rule in `//tools/build_deps/repo`, and
2. [golang.org/x/tools/go/vcs](https://godoc.org/golang.org/x/tools/go/vcs) via rule_go's [fetch_repo](https://github.com/bazelbuild/rules_go/tree/master/go/tools/fetch_repo)

Sidenote: I'll point you to this [rules_go proposal](https://github.com/bazelbuild/rules_go) which recommends the removal
of the go_repository rule in favour of using the official VCS fetch rules.

The reponse to my request for increased VCS system support is justifiably
"contributions are welcome". I submit that contributions are more likely if:

1. (as requested already) the native rules are deprecated and there is only one
place where contributions would need to go, and

2. a small, official location for VCS rules exists, allowing contributors to
do their work without being tied to the much-larger Bazel Git repository.

Therefore, I propose that in addition to deprecating the native Java WORKSPACE
rules that `//tools/build_defs/repo` be moved to its own repository such as
github.com/bazelbuild/vcs_rules. I make both requests at the same time because deprecating and THEN moving the repository would result in multiple migration steps for users coming from the native WORKSPACE VCS rules.

**Comment by kchodorow** [comment]
Agreed that we need this & native rules should be replaced with Skylark versions, but one nit: `--override_repository` works on a repository-by-repository basis and works fine for native rules.

**Comment by pmcalpine** [comment]
> Agreed that we need this & native rules should be replaced with Skylark versions, but one nit: --override_repository ... works fine for native rules.

Does it? What repository would I override to have the native rules replaced with my own?

**Comment by kchodorow** [comment]
@bazel_tools

**Comment by pmcalpine** [comment]
I want to override http_archive itself. I've been able to override the one in @bazel_tools but not the native Java ones. @bazel_tools are the Skylark rules I'd like to see exported to their own rules_vcs repo.

**Comment by ianthehat** [comment]
I thought once before that it would be nice if we added *all* the skylark rule repositories we eventually want to exist (github.com/bazelbuild/rules_java etc), have them be trivial wrappers for now, eg:
```
def java_library(**kwargs):
    return native.java_library(**kwargs)
```
and then eventaully remove the default name bindings so you are force to load these rule sets.
This would allow people to override the native rules if they want, provide a clear path to not having native rules in the future that does not break BUILD files, and put the current native languages closer to being on the same footing as all the other languages.

**Comment by katre** [comment]
@pmcalpine You should be able to write your own http_archive rule in Skylark, then add a `load()` call  in your WORKSPACE file.
```
load(":my_http_archive.bzl", "http_archive")
```
After that, all `http_archive` rules should use your version, not the native one.

**Comment by pmcalpine** [comment]
@katre see my original motivation, I'd like to be able to inject my own implementation of the rules without modifying the WORKSPACE file.

> My motivating objective in this request is to be able to efficiently override
how Bazel fetches dependencies without modifying the WORKSPACE file.
